### DISCUSSION TITLE
Proper way to configure JDBI for spring to work with @Transactional annotations

### DISCUSSION THREAD

**Discussion by d2k2-git** [jdbi/jdbi#2602] [General]
I have a spring boot application which is also using `@Transactional` annotations in the service layer.

**Question 1:**

To make `@Transactional ` work i have to wrap the DataSource either with TranscationAwareDataSourceProxy or use JDBI's SpringConnectionFactory when creating the JDBI bean. Which is the proper way?

**Question 2:**
I saw in same code samples that a PlatformTransactionManager is defined. But it doesn't seem to be required to make the @Transactional annotation work. What is the pupose of this bean? Should i always create an instance of this Bean in my configuration class ?

```java
@Configuration
public class JdbiConfiguration {

    @Bean
    public DataSource dataSource( ) {
        // create any DataSource here...
    }

    /**
     * what is this required for? @Transactional works even without defining this bean
     *
     * @param dataSource
     * @return
     */
    @Bean
    public PlatformTransactionManager platformTransactionManager(DataSource dataSource) {
        return new DataSourceTransactionManager(dataSource);
    }

    @Bean
    public Jdbi jdbi(DataSource dataSource, List<RowMapper<?>> rowMappers) {

        // TransactionAwareDataSourceProxy vs SpringConnectionFactory

        // Jdbi jdbi = Jdbi.create(new TransactionAwareDataSourceProxy(dataSource)); // works for @Transactional
        Jdbi jdbi = Jdbi.create(new SpringConnectionFactory(dataSource));            // works also for @Transactional

        return jdbi;
    }
}
```

**Comment by d2k2-git** [comment]
another thing: i have been debugging my spring boot application and noticed:

1) when i don't define explicitly a PlatformTransactionManager, spring will already define a [JdbcTransactionManager ](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/support/JdbcTransactionManager.html) by default which inherits from DataSourceTransactionManager.

In your test sourcecode i see that jdbi is defining explicitly a DataSourceTransactionManager: https://github.com/jdbi/jdbi/blob/d27dcc7f2ff392c960237fb35ede34499c3e51ca/spring5/src/test/resources/org/jdbi/v3/spring5/test-context.xml#L23C78-L23C106

Why you override it? is it not recommend to use the already created JdbiTransactionManager? might it it lead to unintentional behavior?

**Comment by RezaLabudeBSM** [comment]
Hello,
I'm no expert on JDBI, but know a thing or two about Spring and Spring Boot. The "magic" of Spring Boot is, that it "automagically"adds missing Beans when it find out that they are necessary, but not defined. As Spring Boot saw that you have a datasource defined and that you  use Transactional, it concluded that a Platformtransaction manager is necessary (so when you use Transactional you must also have a PlatformtransactionManager as it is mandatory). Since you did not define it, it created the a JdbcTransactionManager (mind the c, it is NOT a JdbiTransactionManager). It depends on your usecase whether the default Transactionmanager is enought or you need to configure and adapt it to your scenario. If you define it by yourself, you know what you get and you are in full control and responsibility, if you let Spring Boot decide, you might end up with something you do not want or that does not work as you expect it, as it is intended as a good default fallback. 
I hope this helps and answers your questions.
best regards 
Reza
